tpcc/new_order.go

```diff
@@ -155,88 +166,165 @@ func (n newOrder) run(db *sql.DB, wID int) (interface{}, error) {
-				VALUES ($1, $2, $3)`,
-				d.oID, d.dID, d.wID); err != nil {
-				return err
+			// 2.4.2.2: For each o_ol_cnt item in the order, query the relevant item
+			// row, update the stock row to account for the order, and insert a new
+			// line into the order_line table to reflect the item on the order.
+			itemIDs := make([]interface{}, d.oOlCnt)
+			itemIDPls := make([]string, d.oOlCnt)
+			for i, item := range d.items {
+				itemIDs[i] = item.olIID
+				itemIDPls[i] = fmt.Sprintf("$%d", i+1)
 			}
-
-			selectItem, err := tx.Prepare(`SELECT i_price, i_name, i_data FROM item WHERE i_id=$1`)
+			rows, err := tx.Query(fmt.Sprintf(`
+				SELECT i_price, i_name, i_data 
+				FROM item 
+				WHERE i_id IN (%s)`, strings.Join(itemIDPls, ", ")),
+				itemIDs...,
+			)
 			if err != nil {
 				return err
 			}
-			updateStock, err := tx.Prepare(fmt.Sprintf(`
-			UPDATE stock
-			SET (s_quantity, s_ytd, s_order_cnt, s_remote_cnt) =
-				(CASE s_quantity >= $1 + 10 WHEN true THEN s_quantity-$1 ELSE (s_quantity-$1)+91 END,
-				 s_ytd + $1,
-				 s_order_cnt + 1,
-				 s_remote_cnt + (CASE $2::bool WHEN true THEN 1 ELSE 0 END))
-			WHERE s_i_id=$3 AND s_w_id=$4
-			RETURNING s_dist_%02d, s_data`, d.dID))
-			if err != nil {
+			iDatas := make([]string, d.oOlCnt)
+			for i := 0; rows.Next(); i++ {
+				item := &d.items[i]
+				iData := &iDatas[i]
+				err = rows.Scan(&item.iPrice, &item.iName, iData)
+				if err != nil {
+					rows.Close()
+					return err
+				}
+			}
+			if err := rows.Err(); err != nil {
 				return err
 			}
-			insertOrderLine, err := tx.Prepare(`
-			INSERT INTO order_line(ol_o_id, ol_d_id, ol_w_id, ol_number, ol_i_id, ol_supply_w_id, ol_quantity, ol_amount, ol_dist_info)
-			VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)`)
+			rows.Close()
+			if rollback {
+				// 2.4.2.3: roll back when we're expecting a rollback due to
+				// simulated user error (invalid item id) and we actually
+				// can't find the item. The spec requires us to actually go
+				// to the database for this, even though we know earlier
+				// that the item has an invalid number.
+				return errSimulated
+			}
+
+			stockIDs := make([]interface{}, 2*d.oOlCnt)
+			stockIDPls := make([]string, d.oOlCnt)
+			for i, item := range d.items {
+				stockIDs[2*i] = item.olIID
+				stockIDs[2*i+1] = item.olSupplyWID
+				stockIDPls[i] = fmt.Sprintf("($%d, $%d)", 2*i+1, 2*i+2)
+			}
+			rows, err = tx.Query(fmt.Sprintf(`
+				SELECT s_quantity, s_ytd, s_order_cnt, s_remote_cnt, s_data, s_dist_%02d
+				FROM stock
+				WHERE (s_i_id, s_w_id) IN (%s)`,
+				d.dID, strings.Join(stockIDPls, ", ")),
+				stockIDs...,
+			)
 			if err != nil {
 				return err
 			}
+			distInfos := make([]string, d.oOlCnt)
+			sQuantityUpdateCases := make([]string, d.oOlCnt)
+			sYtdUpdateCases := make([]string, d.oOlCnt)
+			sOrderCntUpdateCases := make([]string, d.oOlCnt)
+			sRemoteCntUpdateCases := make([]string, d.oOlCnt)
+			for i := 0; rows.Next(); i++ {
```

Comment:
I'm nervous that this is broken because the select doesn't have an order by... don't we need the results to be in the same order that we asked for from d.items?

Reply:
Yeah, you're completely right. I meant to fix this when I made the change to ensure that all items are unique.